### PR TITLE
#363 Wire Up Project Edit Form

### LINKED ISSUE #363: Projects - Wire Up Project Edit Form
## Prerequisites 

- [x] #361
- [x] #370
- [x] #367
- [x] #466

## Description
This ticket is to complete the project edit feature by wiring up the front-end form to the back-end endpoint.  

## Proposed Solution

Implement the new reusable editable list component created in #466 / #518. Read through that ticket and pull request pair to ensure you understand how to implement it.

https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/3e46727a554bb5882a944ca5bf4e3d113b9f9a9a/src/components/shared/editable-text-input-list/editable-text-input-list.tsx#L15

Gather together data from the front-end form into state management storage in the container component. Then facilitate packaging and sending the data to the back-end by:

- [ ] writing the api function
- [ ] writing the custom hook (using `useMutation`)
- [ ] implementing the custom hook  (using the payload schema in the `utils` package)

Ensure all aspects of the project edit feature are present and functioning. Verify by tracing data from user input through the full stack to database ingestion. Implement / fix / write tests for any of the following elements in the trace through the tech stack that are not behaving as expected.

- [ ] Delete buttons
- [ ] Add buttons
- [ ] Save/Cancel buttons
- [ ] Form input fields
- [ ] Mutation hook for submitting data
- [ ] Api function sending data to backend
- [ ] Api route string
- [ ] Api route function
- [ ] Payload validation
- [ ] Database insertion
- [ ] Api response

### PR DESCRIPTION
**Pull request by anthonybernardi**
## Changes

Added the wiring up functionality as specified in the tickets. Not done with testing yet but want review on what I have so far. Manual testing shows that it works right.

## Test Cases

have not tested yet

## To Do

- [x] testing

## Checklist

- [x] All commits are tagged with the ticket number
- [x] No linting errors
- [x] No newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (if applicable)
- [x] Remove any not-applicable sections
- [x] Assign the PR to yourself
- [x] PR is linked to the ticket
- [x] No `package-lock.json` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack

Closes #363

### CODE REVIEW

File: src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx [hunk 1]
```diff
@@ -41,10 +53,28 @@ const ProjectEditContainer: React.FC<EditFormContainerProps> = ({ wbsNum, proj,
...
 
-  const [goals, setGoals] = useState(proj!.goals.map((goal) => goal.detail));
-  const [features, setFeatures] = useState(proj!.features.map((feature) => feature.detail));
-  const [otherConstraints, setOther] = useState(
-    proj!.otherConstraints.map((constraint) => constraint.detail)
+  const { mutateAsync } = useEditSingleProject();
+
+  const [goals, setGoals] = useState<{ id?: number; detail: string }[]>(
+    proj!.goals
+      .filter((goal) => !goal.dateDeleted)
```
Comment (jamescd18) [inline src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx:60]: I guess this should fix the issue, i think i might have submitted a bug ticket for it? I like it

File: src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx [hunk 2]
```diff
@@ -41,10 +53,28 @@ const ProjectEditContainer: React.FC<EditFormContainerProps> = ({ wbsNum, proj,
...
+      .filter((feature) => !feature.dateDeleted)
+      .map((feature) => {
+        return { id: feature.id, detail: feature.detail };
+      })
+  );
+  const [otherConstraints, setOther] = useState<{ id?: number; detail: string }[]>(
+    proj!.otherConstraints
+      .filter((constraint) => !constraint.dateDeleted)
+      .map((constraint) => {
+        return { id: constraint.id, detail: constraint.detail };
```
Comment (jamescd18) [inline src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx:76]: Abstraction? We do this 3 times? Just thinking out loud

**Review by jamescd18 — COMMENTED** [review]
LGTM so far. Will try and do manual testing. Looking forward to seeing this tested a bit and then merged to go!

File: src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx (hunk 2 above)
Comment (anthonybernardi) [reply src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx]: true!

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Looks like errors aren't being handled?

File: src/backend/functions/projects-edit.ts [hunk 3]
```diff
@@ -88,6 +88,8 @@ export const editProject: Handler = async ({ body }, _context) => {
     return buildClientFailureResponse('Project with given wbsElementId does not exist!');
```
Comment (jamescd18) [inline src/backend/functions/projects-edit.ts:88]: not using wbsElementId anymore, right?

**Review by jamescd18 — CHANGES_REQUESTED** [review]

File: src/backend/functions/projects-edit.ts (hunk 3 above)
Comment (anthonybernardi) [reply src/backend/functions/projects-edit.ts]: tru

File: src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx [hunk 4]
```diff
@@ -41,10 +63,16 @@ const ProjectEditContainer: React.FC<EditFormContainerProps> = ({ wbsNum, proj,
   const updateBom = (url: string | undefined) => setBom(url);
   const updateGDrive = (url: string | undefined) => setGDrive(url);
 
-  const [goals, setGoals] = useState(proj!.goals.map((goal) => goal.detail));
-  const [features, setFeatures] = useState(proj!.features.map((feature) => feature.detail));
-  const [otherConstraints, setOther] = useState(
-    proj!.otherConstraints.map((constraint) => constraint.detail)
+  const { mutateAsync } = useEditSingleProject();
```
[the remark this answers is not in the document]
Comment (anthonybernardi) [reply src/components/projects/project-edit-form/project-edit-container/project-edit-container.tsx:66]: done

**Review by jamescd18 — APPROVED** [review]
LGTM

**Pull request merged**